### PR TITLE
[EC-147] Decide on what to do with MPTNodeStorage and NodeStorage

### PR DESCRIPTION
**Pull request by adamsmo**
this PR removes mpt nodes from PV63

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala [hunk 1]
```diff
@@ -167,55 +174,6 @@ object PV63 {
     }
   }
 
-  case class MptBranch(children: Seq[Either[MptHash, MptNode]], value: ByteString) extends MptNode {
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala]: :rocket:

File: src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala [hunk 2]
```diff
@@ -108,8 +107,7 @@ object MerklePatriciaTrie {
 trait NodesKeyValueStorage extends SimpleMap[NodeHash, NodeEncoded, NodesKeyValueStorage]
 
 class MerklePatriciaTrie[K, V] private (private val rootHash: Option[Array[Byte]],
-  val nodeStorage: NodesKeyValueStorage,
-  private val hashFn: HashFn)
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala]: :+1:

File: src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala [hunk 3]
```diff
@@ -15,46 +14,46 @@ object MerklePatriciaTrie {
...
+    toUpdateInStorage: Seq[MptNode] = Nil)
 
-  private case class NodeRemoveResult(hasChanged: Boolean, newNode: Option[Node],
-    toDeleteFromStorage: Seq[Node] = Nil,
-    toUpdateInStorage: Seq[Node] = Nil)
+  private case class NodeRemoveResult(hasChanged: Boolean, newNode: Option[MptNode],
+    toDeleteFromStorage: Seq[MptNode] = Nil,
+    toUpdateInStorage: Seq[MptNode] = Nil)
 
   type HashFn = Array[Byte] => Array[Byte]
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala]: I think we can also remove this one

File: src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala [hunk 4]
```diff
@@ -15,46 +14,46 @@ object MerklePatriciaTrie {
...
+  def apply[K, V](rootHash: Array[Byte], source: NodesKeyValueStorage)
     (implicit kSerializer: ByteArrayEncoder[K], vSerializer: ByteArraySerializable[V])
   : MerklePatriciaTrie[K, V] = {
-    if (calculateEmptyRootHash(hashFn) sameElements rootHash) MerklePatriciaTrie(source, hashFn)
-    else new MerklePatriciaTrie[K, V](Some(rootHash), source, hashFn)(kSerializer, vSerializer)
+    if (calculateEmptyRootHash(Node.hashFn) sameElements rootHash) MerklePatriciaTrie(source)
+    else new MerklePatriciaTrie[K, V](Some(rootHash), source)(kSerializer, vSerializer)
   }
 
   def calculateEmptyRootHash(hashFn: HashFn): Array[Byte] = hashFn(encodeRLP(Array.emptyByteArray))
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala]: As `hashFn` is no longer configurable, imo the parameter is no longer needed

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala [hunk 5]
```diff
@@ -74,19 +75,24 @@ object PV63 {
     implicit class MptNodeEnc(obj: MptNode) extends RLPSerializable {
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala]: Isn't this encoder the same as this one https://github.com/input-output-hk/mantis/pull/291/files#diff-081f56715a2b593d7116a59f2b2a251eR80 ?

File: src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala [hunk 6]
```diff
@@ -118,32 +114,32 @@ class MerklePatriciaTrie[K, V] private (private val rootHash: Option[Array[Byte]
   /**
     * Implicits
     */
-  private[mpt] implicit val nodeDec = new RLPDecoder[Node] {
-    override def decode(rlp: RLPEncodeable): Node = rlp match {
+  private[mpt] implicit val nodeDec = new RLPDecoder[MptNode] {
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala]: As there is no more custom `hashFn` I think we can move this away from `class MerklePatriciaTrie`

File: src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala [hunk 7]
```diff
@@ -15,46 +14,44 @@ object MerklePatriciaTrie {
...
     (implicit kSerializer: ByteArrayEncoder[K], vSerializer: ByteArraySerializable[V])
   : MerklePatriciaTrie[K, V] = {
-    if (calculateEmptyRootHash(hashFn) sameElements rootHash) MerklePatriciaTrie(source, hashFn)
-    else new MerklePatriciaTrie[K, V](Some(rootHash), source, hashFn)(kSerializer, vSerializer)
+    if (calculateEmptyRootHash() sameElements rootHash) MerklePatriciaTrie(source)
+    else new MerklePatriciaTrie[K, V](Some(rootHash), source)(kSerializer, vSerializer)
   }
 
-  def calculateEmptyRootHash(hashFn: HashFn): Array[Byte] = hashFn(encodeRLP(Array.emptyByteArray))
+  def calculateEmptyRootHash(): Array[Byte] = Node.hashFn(encodeRLP(Array.emptyByteArray))
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala]: Minor detail: Maybe we could replace this with a `val` as it never changes?

File: src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala [hunk 8]
```diff
@@ -241,37 +236,37 @@ class MerklePatriciaTrie[K, V] private (private val rootHash: Option[Array[Byte]
   }
 
   @tailrec
-  private def get(node: Node, searchKey: Array[Byte]): Option[Array[Byte]] = node match {
-    case LeafNode(key, value, _) =>
-      if (key sameElements searchKey) Some(value) else None
-    case extNode@ExtensionNode(sharedKey, _, _) =>
+  private def get(node: MptNode, searchKey: Array[Byte]): Option[Array[Byte]] = node match {
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala]: shouldn't it be better to change `searchKey` to ByteArray and the result to `Option[ByteString]` ?

File: src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala [hunk 9]
```diff
@@ -118,32 +114,32 @@ class MerklePatriciaTrie[K, V] private (private val rootHash: Option[Array[Byte]
...
           }
-          ExtensionNode(key, next, hashFn)
+          ExtensionNode(ByteString(key), next)
         }
       case _ => throw new RuntimeException("Invalid Node")
     }
   }
 
-  private lazy val EmptyTrieHash = hashFn(encodeRLP(Array.emptyByteArray))
+  private lazy val EmptyTrieHash = Node.hashFn(encodeRLP(Array.emptyByteArray))
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala]: Could it be that this is duplicated with `calculateEmptyRootHash`? Maybe we could leave just `calculateEmptyRootHash`

File: src/main/scala/io/iohk/ethereum/mpt/Node.scala [hunk 10]
```diff
@@ -1,63 +1,71 @@
...
 }
 
-private[mpt] case class LeafNode(key: Array[Byte], value: Array[Byte], hashFn: HashFn) extends Node
+object Node {
+  val hashFn: (Array[Byte]) => Array[Byte] = (input: Array[Byte]) => crypto.kec256(input)
+}
+
+case class LeafNode(key: ByteString, value: ByteString) extends MptNode {
+  import AccountImplicits._
+  def getAccount: Account = value.toArray[Byte].toAccount
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/mpt/Node.scala]: As the MPT is used for other things apart from accounts, maybe we should move this function somewhere else?

File: src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala (hunk 8 above)
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala]: I did not want to change mpt implementation too much, maybe we can do this refactoring in separate PR?

Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala]: :+1: makes sense

File: src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala (hunk 2 above)
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/mpt/MerklePatriciaTrie.scala]: :+1:

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala (hunk 5 above)
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala]: encoders merged

**Comment by adamsmo** [comment]
@AlanVerbner / @ntallar updated, ready for review

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncNodesRequestHandler.scala [hunk 11]
```diff
@@ -76,11 +78,14 @@ class FastSyncNodesRequestHandler(
   }
 
   private def handleMptNode(mptNode: MptNode): Seq[HashType] = mptNode match {
-    case n: MptLeaf =>
-      val evm = n.getAccount.codeHash
-      val storage = n.getAccount.storageRoot
+    case n: LeafNode =>
+      import AccountImplicits._
+      val account = n.value.toArray[Byte].toAccount
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncNodesRequestHandler.scala]: Should we maybe handle the case that this conversion were to fail? As the node was received over the network it could be the case that it's invalid and doesn't contain an account, however it's up to discussion.

File: src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala [hunk 12]
```diff
@@ -74,19 +74,24 @@ object PV63 {
...
 
       def toRLPEncodable: RLPEncodeable = obj match {
```
Comment (ntallar) [inline src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala]: Should we maybe also replace this encoder with `MerklePatriciaTrie.nodeDec.encode` so as to not have the encoder duplicated?

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncNodesRequestHandler.scala [hunk 13]
```diff
@@ -76,49 +79,56 @@ class FastSyncNodesRequestHandler(
...
 
   private def handleMptNode(mptNode: MptNode): Seq[HashType] = mptNode match {
-    case n: MptLeaf =>
-      val evm = n.getAccount.codeHash
-      val storage = n.getAccount.storageRoot
+    case n: LeafNode =>
+      import AccountImplicits._
+      //if this fails it means that we have leaf node which is part of MPT that do not stores account
+      //we verify if node is paert of the tree by checking its hash before we call handleMptNode() in line 44
+      val account = Try(n.value.toArray[Byte].toAccount)
```
Comment (KonradStaniec) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncNodesRequestHandler.scala]: minor readability issue (it was hard for me to get intention here). Imo pattern match on try would be clearer, without all these transformations. Something like:
```  
case Success(acc) => Some(acc)
case Failure(e) =>
    log.debug(s"Leaf node without account, error while trying to decode account ${e.getMessage}")
    None
```

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncNodesRequestHandler.scala [hunk 14]
```diff
@@ -76,49 +79,56 @@ class FastSyncNodesRequestHandler(
   }
 
   private def handleMptNode(mptNode: MptNode): Seq[HashType] = mptNode match {
```
Comment (KonradStaniec) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncNodesRequestHandler.scala]: This function is almost the same as `handleContractMptNode`, especially `ExtensionNode` and `BranchNode` cases, maybe w can unify it somehow.
I was thinking something along these lines for extension node in both of them:
``` 
private def handleExtensionNode(n: ExtensionNode)(f: ByteString => Seq[HashType]): Seq[HashType] = {
    saveNodeFn(ByteString(n.hash), n.toBytes)
    n.next.fold(f, _ => Nil)
  }
```

File: src/main/scala/io/iohk/ethereum/mpt/Node.scala [hunk 15]
```diff
@@ -1,63 +1,68 @@
...
-private[mpt] sealed trait Node {
+sealed trait MptNode {
 
   import MerklePatriciaTrie._
 
-  protected val hashFn: HashFn
+  lazy val encode: Array[Byte] = encodeRLP[MptNode](this)
 
-  lazy val encode: Array[Byte] = encodeRLP[Node](this)
+  lazy val hash: Array[Byte] = Node.hashFn(encode)
```
Comment (LukasGasior1) [inline src/main/scala/io/iohk/ethereum/mpt/Node.scala]: how about ByteString? (it seems to be converted to it anyway in quite a few places)

File: src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncNodesRequestHandler.scala (hunk 14 above)
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncNodesRequestHandler.scala]: I am not sure if those 2 lines are worth this

File: src/it/scala/io/iohk/ethereum/txExecTest/util/FixtureProvider.scala [hunk 16]
```diff
@@ -23,14 +22,14 @@ import scala.util.Try
...
   case class Fixture(
-    blockByNumber: Map[BigInt, Block],
-    blockByHash: Map[ByteString, Block],
-    blockHeaders: Map[ByteString, BlockHeader],
-    blockBodies: Map[ByteString, BlockBody],
-    receipts: Map[ByteString, Seq[Receipt]],
-    stateMpt: Map[ByteString, MptNode],
-    contractMpts: Map[ByteString, MptNode],
-    evmCode: Map[ByteString, ByteString]
+                      blockByNumber: Map[BigInt, Block],
```
Comment (adamsmo) [inline src/it/scala/io/iohk/ethereum/txExecTest/util/FixtureProvider.scala]: yes

File: src/main/scala/io/iohk/ethereum/mpt/Node.scala (hunk 15 above)
Comment (adamsmo) [inline src/main/scala/io/iohk/ethereum/mpt/Node.scala]: no, it is not that simple, I was trying to change this but MPT implementation relays on this to bey byte array, so it is true that in some cases it is converted to Bytestring but in many places it is used as Array[Byte], I was trying to change this to Bytestring but than MPT was failing so I decided to leave it as is
